### ISSUE TITLE
Inaccuracy in the cross correlation for small batch sizes (<32)

### ISSUE THREAD

**Issue by JohannesK14** [yaohungt/Barlow-Twins-HSIC#6]
Hey,

first, thanks for sharing your research and code!

### TL;DR
**Your code uses torch.std which uses Bessel's correction by default, therefore inhibiting that the values on the diagonal reach 1.**

While working with it, I noticed some small inaccuracy for in the calculation of the cross-correlation matrix.

Opposed to original implementation, which uses [BatchNorm1d](https://pytorch.org/docs/stable/generated/torch.nn.BatchNorm1d.html) you implemented the normalization with:

https://github.com/yaohungt/Barlow-Twins-HSIC/blob/a30baba4d2d3dcdf85a4ccebe57902f0d827c1ed/main.py#L36-L38

I implemented a small test with two identical vectors coming from the projection head and was therefore expecting straight ones on the diagonal. But as you can see from my attached code, for a batch size of < 32 (here 8), the values on the diagonal can't get bigger than 0.75. I found that torch.std uses Bessel's correction by default. When this flag is set to false, the numbers match with the original implementation.

I think there is no practical difference for batch sizes > 32, which is also the smallest batch size you presented in your paper, I think.

    import torch
    from torch import nn
    
    batch_size = 4
    size_z = 128
    
    torch.manual_seed(1234)
    z1 = torch.randn(batch_size, size_z)
    z2 = z1.clone()

    # your implementation
    z1_norm = (z1 - torch.mean(z1, dim=0)) / torch.std(z1, dim=0)
    z2_norm = (z2 - torch.mean(z2, dim=0)) / torch.std(z2, dim=0)
    cross_corr = torch.matmul(z1_norm.T, z2_norm) / batch_size
    
    print( cross_corr[0:5,0:5] )
    
    # tensor([[ 0.7500, -0.1065, -0.0837,  0.0630,  0.3664],
    #        [-0.1065,  0.7500, -0.2283, -0.3708, -0.5607],
    #        [-0.0837, -0.2283,  0.7500, -0.5013, -0.2554],
    #        [ 0.0630, -0.3708, -0.5013,  0.7500,  0.6334],
    #        [ 0.3664, -0.5607, -0.2554,  0.6334,  0.7500]])

    # original implementation
    bn = nn.BatchNorm1d(size_z, affine=False)
    z1_norm = bn(z1)
    z2_norm = bn(z2)
    cross_corr = z1_norm.T @ z2_norm / batch_size
    
    print( cross_corr[0:5,0:5] )
    
    # tensor([[ 1.0000, -0.1420, -0.1116,  0.0840,  0.4885],
    #        [-0.1420,  1.0000, -0.3043, -0.4944, -0.7476],
    #        [-0.1116, -0.3043,  1.0000, -0.6683, -0.3405],
    #        [ 0.0840, -0.4944, -0.6683,  1.0000,  0.8445],
    #        [ 0.4885, -0.7476, -0.3405,  0.8445,  1.0000]])
    
    # corrected code (without Bessel’s correction for the calculation of the standard deviation)
    z1_norm = (z1 - torch.mean(z1, dim=0)) / torch.std(z1, dim=0, unbiased=False)
    z2_norm = (z2 - torch.mean(z2, dim=0)) / torch.std(z2, dim=0, unbiased=False)
    cross_corr = torch.matmul(z1_norm.T, z2_norm) / batch_size
    
    print( cross_corr[0:5,0:5] )
    
    # tensor([[ 1.0000, -0.1421, -0.1116,  0.0840,  0.4885],
    #         [-0.1421,  1.0000, -0.3043, -0.4944, -0.7476],
    #         [-0.1116, -0.3043,  1.0000, -0.6683, -0.3405],
    #         [ 0.0840, -0.4944, -0.6683,  1.0000,  0.8446],
    #         [ 0.4885, -0.7476, -0.3405,  0.8446,  1.0000]])
